nservicebus/messaging/dynamic-dispatch-and-routing.md

```diff
@@ -0,0 +1,40 @@
+---
+title: Dynamic dispatch and polymorphic routing
+summary: Implement polymorphic routing with NServiceBus-supported dynamic dispatch to route messages based on their types
+component: core
+reviewed: 2024-04-04
+related:
+ - nservicebus/messaging/messages-as-interfaces
+---
+
+NServiceBus makes it easy to support dynamic dispatch and polymorphic message routing scenarios. This technique makes systems more flexible and maintainable by encouraging messages and handlers to be decoupled from each other. It's useful for evolving systems by allowing the addition of new features without affecting existing functionality. It's best described using an example.
+
+## Using NServiceBus for dynamic dispatch
+
+Consider the following handler stubs for a `UserCreated` message which inherits from the `IEvent` stub interface:
+
+snippet: DynamicDispatchHandlerStubs
+
+When a message of type `UserCreated` is published, both handlers will be invoked independently which in itself is useful. However, this idea becomes more powerful as the system evolves and more functionality is added.
+
+Meanwhile, imagine there is a separate process that must happen if a user is created as part of an advertising campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
+
+snippet: DynamicDispatchEvolution
+
+Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers. This is the essence of dynamic dispatch: the system decides at runtime which handlers to invoke based on the message type that is being processed.
```

Comment:
```suggestion
With the introduction of this new code, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers. This is the essence of dynamic dispatch: the system decides at runtime which handlers to invoke based on the message type that is being processed.
```